superset-frontend/plugins/plugin-chart-period-over-period-kpi/src/plugin/buildQuery.ts

```diff
@@ -0,0 +1,283 @@
+      (delta: string, unit: string) =>
+        moment().subtract(delta, unit as moment.unitOfTime.DurationConstructor),
+    ],
+    [
+      /^next\s+([0-9]+)\s+(second|minute|hour|day|week|month|year)s?$/i,
+      (delta: string, unit: string) =>
+        moment().add(delta, unit as moment.unitOfTime.DurationConstructor),
+    ],
+    [
+      // eslint-disable-next-line no-useless-escape
+      /DATEADD\(DATETIME\("([^"]+)"\),\s*(-?\d+),\s*([^\)]+)\)/i,
+      (timePart: string, delta: string, unit: string) => {
+        if (timePart === 'now') {
+          return moment().add(
+            delta,
+            unit as moment.unitOfTime.DurationConstructor,
+          );
+        }
+        if (moment(timePart.toUpperCase(), true).isValid()) {
+          return moment(timePart).add(
+            delta,
+            unit as moment.unitOfTime.DurationConstructor,
+          );
+        }
+        return moment();
+      },
+    ],
+  ];
+
+  const sinceAndUntilPartition = modTimeRange
+    .split(separator, 2)
+    .map(part => part.trim());
+
+  const sinceAndUntil: (Moment | null)[] = sinceAndUntilPartition.map(part => {
+    if (!part) {
+      return null;
+    }
+
+    let transformedValue: Moment | null = null;
+    // Matching time_range_lookup
+    const matched = timeRangeLookup.some(([pattern, fn]) => {
+      const result = part.match(pattern);
+      if (result) {
+        transformedValue = fn(...result.slice(1));
+        return true;
+      }
+
+      if (part === 'today') {
+        transformedValue = moment().startOf('day');
+        return true;
+      }
+
+      if (part === 'now') {
+        transformedValue = moment();
+        return true;
+      }
+      return false;
+    });
+
+    if (matched && transformedValue !== null) {
+      // Handle the transformed value
+    } else {
+      // Handle the case when there was no match
+      transformedValue = moment(`${part}`);
+    }
+
+    return transformedValue;
+  });
+
+  const [_since, _until] = sinceAndUntil;
+
+  if (_since && _until && _since.isAfter(_until)) {
+    throw new Error('From date cannot be larger than to date');
+  }
+
+  return [_since, _until];
+}
+
+function calculatePrev(startDate: Moment | null, endDate: Moment | null, calcType: String) {
+  
+  if (!startDate || !endDate){
+    return [null, null]
+  }
+  
+  const daysBetween = endDate.diff(startDate, 'days');
+
+  let startDatePrev = moment();
+  let endDatePrev = moment();
+  if (calcType === 'y') {
+    startDatePrev = startDate.subtract(1, 'year');
+    endDatePrev = endDate.subtract(1, 'year');
+  } else if (calcType === 'w') {
+    startDatePrev = startDate.subtract(1, 'week');
+    endDatePrev = endDate.subtract(1, 'week');
+  } else if (calcType === 'm') {
+    startDatePrev = startDate.subtract(1, 'month');
+    endDatePrev = endDate.subtract(1, 'month');
+  } else if (calcType === 'r') {
+    startDatePrev = startDate.clone().subtract(daysBetween.valueOf(), 'day');
```

Comment:
`diff` should already return a number so `valueOf` might be redundant here

Reply:
Do we need `clone` here?